File: triples.php
<?php
/* Basic script for handling Browsing and Modifiation of Triples
 *
 * author   : Johann-Mattis List
 * email    : mattis.list@lingulist.de
 * created  : 2014-08-31 22:09
 * modified : 2014-09-07 17:33
 *
 */
?>
<?php
header('HTTP/1.1 200 OK');
header('Content-Type: text/plain; charset=utf-8');
header('Content-Disposition: attachment; filename="triples.tsv"');
$now = date('Y-m-d H:i:s');
if(isset($_SERVER['REMOTE_USER'])) {
  $user = $_SERVER['REMOTE_USER'];
}
else {
  $user = 'unknown'; 
}

if (isset($_GET['remote_dbase'])) {
  $dsn = "sqlite:".$_GET['remote_dbase'];
}
else {
  $dsn = "sqlite:triples.sqlite3";
}
$con = new PDO ($dsn);
/* this is our sorter function that handles the order of
 * columns in the data 
 */
function sortMyCols($valA,$valB) {
  $sorter = array(
    "DOCULECT" => 1,
    "CONCEPT" => 2,
    "IPA" => 3,
    "TOKENS" => 4,
    "CLUSTERID" => 5,
    "ALIGNMENT" => 6
  );
  if(array_key_exists($valA, $sorter)) {
    $valA = $sorter[$valA];
  }
  else {
    $valA = 10;
  }
  if(array_key_exists($valB, $sorter)) {
    $valB = $sorter[$valB];
  }
  else {
    $valB = 10;
  }
  if($valA < $valB) {return -1; }
  if($valB < $valA) {return 1; }
  if($valA == $valB) {return 0; }
}
/* if site is called with keyword "tables", return all tables, each in 
 * one line 
 */
if(isset($_GET['tables'])) {

  $query = $con->query('select name from sqlite_master where name != "backup";');
  $data = $query->fetchAll(PDO::FETCH_COLUMN, 0);
  foreach ($data as $table) {
    echo $table."\n";
  }
}
/* return most recent edits, if this is chosen */
else if (isset($_GET['date'])) {
  $query = $con->query('select ID,COL from backup where FILE="' .
    $_GET['file'] . '" and datetime(DATE) > datetime("' . $_GET['date'] .
    '") group by ID,COL;');
  $data = $query->fetchAll();
  foreach ($data as $line) {
    $query = $con->query('select VAL from ' . $_GET['file'] . ' where ID=' .
      $line['ID'] . ' and COL like "' . $line['COL'] .'";');
    $tmp = $query->fetch();
    echo $line['ID'] . "\t" . $line["COL"] . "\t" . $tmp["VAL"] . "\n";
  }
}
else if (isset($_GET['new_id'])) {
  /* get all indices */
  /* start with the case that an explicit ID is chosen */
  if ($_GET['new_id'] == '') {
    $sth = $con->prepare('select DISTINCT ID from ' . $_GET['file'] . ';');
    $sth->execute();
    $idxA = $sth->fetchAll(PDO::FETCH_COLUMN, 0);
  
    /* do the same for backup file */
    $sth = $con->prepare('select DISTINCT ID from backup where file == "'. $_GET['file']. '";');
    $sth->execute();
    $idxB = $sth->fetchAll(PDO::FETCH_COLUMN, 0);

    $maxA = max($idxA);
    $maxB = max($idxB);
    if ($maxA >= $maxB) {
      echo $maxA + 1;
    }
    else {
      echo $maxB + 1;
    }
  }
  /* yet another possibility is that the new id is supposed to be a cognate id or the like */
  else {
    $sth = $con->prepare('select DISTINCT VAL from '.$_GET['file'] .' where COL == "'.$_GET['new_id'].'" order by COL;');
    $sth->execute();
    $idxs = $sth->fetchAll(PDO::FETCH_COLUMN, 0);
    for ($i=0;$i<=count($idxs)+1;$i++) {
      if (!in_array($i, $idxs)) {
	echo $i;
	break;
      }
    }
  } 
}

else if(isset($_GET['file'])) {
  
  /* we make some simple solution here: if columns are passed from the 
   * get-line, we modify the query, if not, we represent the modification
   * as an empty string */
  if (!isset($_GET['columns'])) {
    $where = '';
  }
  else {
    $cols = explode(",",$_GET['columns']);
    $col_string = '("'.implode('","',$cols).'")';
    $where = ' where COL in '.$col_string;
  }

  /* set up array for ids we want to have included */
  $sth = $con->prepare('select DISTINCT ID from '.$_GET['file'].';');
  $sth->execute();
  $aids = $sth->fetchAll(PDO::FETCH_COLUMN, 0);
  $array_count = count($aids);
  $dids = $aids;
  $cids = $aids;
  $idxs = array();


  /* similarly, we do the same for doculects and concepts, but we do not
   * combine the queries but rather use them distinctely to narrow down
   * the number of possible ids to be sent off */
  
  /* if doculects are passed, we need to make a pre-selection of possible ids */
  if (isset($_GET['doculects'])) {
    $doc_string = '("'.implode('","',explode(',',$_GET['doculects'])).'")';

    $sth = $con->prepare('select DISTINCT ID from ' . $_GET['file'] .
      ' where VAL in '.$doc_string. ';');
    $sth->execute();
    $dids = $sth->fetchAll(PDO::FETCH_COLUMN,0);
  }

  /* if concepts are passed, we need to further sort by concept selection */
  if (isset($_GET['concepts'])) {
    $con_string = '("'.implode('","',explode(',',$_GET['concepts'])).'")';

    $sth = $con->prepare('select DISTINCT ID from ' . $_GET['file'] .
      ' where VAL in '.$con_string.';');
    $sth->execute();
    $cids = $sth->fetchAll(PDO::FETCH_COLUMN,0);
  }

  $idxs = array_intersect($dids, $cids);
  
  /* get all columns */
  $sth = $con->prepare('select DISTINCT COL from ' . $_GET['file'] . $where. ';');
  $sth->execute();
  $cols = $sth->fetchAll(PDO::FETCH_COLUMN, 0);
  usort($cols, "sortMyCols");

  echo "ID";
  foreach($cols as $col)
  {
    echo "\t".$col;
  }
  echo "\n#\n";

  /* create text */
  $text = "";

  /* $data stores the data in json-like form */
  $data = array();

  $qstring = 'select * from '.$_GET['file'].$where.' and VAL!="-" and VAL!="" and ID in ('.implode(",",$idxs).');';
  $query = $con->query($qstring);
  $results = $query->fetchAll();
  foreach ($results as $entry) {
    try {
      $data[$entry['ID']][$entry['COL']] = $entry['VAL'];
    }
    catch (Exception $e) {
      $data[$entry['ID']] = array($entry['COL'] => $entry['VAL']);
    }
  }

  /* iterate over array and assign all columns */
  foreach ($idxs as $idx) {
    echo $idx;
    foreach ($cols as $col) {
      if (isset($data[$idx][$col])) {
        echo "\t" . $data[$idx][$col];
      }
      else {
        echo "\t";
      }
    }
    echo "\n";
  }
}
/* check the history of edits, if this option is chosen */
else if (isset($_GET['history'])) {
  $query = $con->query('select * from backup;');
  $data = $query->fetchAll();
  foreach ($data as $line) {
    echo $line['FILE'] . "\t" . $line["ID"] . "\t" . $line["COL"] . "\t" .
      $line["VAL"] . "\t" . $line["DATE"] . "\t" .$line["user"] . "\n";
  }
}
else
{
  echo "no parameters specified by user " . $user;
}
?>

File: triples.py
#!/usr/bin/python2.6
import cgi
import sqlite3
import datetime

print "Content-type: text/plain; charset=utf-8"

# get the args of the url, convert nasty field storage to plain dictionary,
# there is probably a better solution, but this works for the moment
tmp_args = cgi.FieldStorage()
args = {}
for arg in tmp_args:
    args[arg] = tmp_args[arg].value

# check for dbase arg and switch the database in case an argument is provided
if 'remote_dbase' in args:
    dbpath = args['remote_dbase']+'.sqlite3' if not \
            args['remote_dbase'].endswith('sqlite3') else \
            args['remote_dbase']
else:
    dbpath = 'triples.sqlite3'

# connect to the sqlite database
db = sqlite3.connect(dbpath)
cursor = db.cursor()

# load the table if this is specified
if 'tables' in args:
    print 
    for line in cursor.execute(
    'select name from sqlite_master where name != "backup";'):
        print line[0]
# return most recent edits in the data
elif 'date' in args:
    print
    cursor.execute(
            'select ID,COL from backup where FILE="'+args['file']+'"'+\
                    ' and datetime(DATE) > datetime('+args['date']+')'+\
                    ' group by ID,COL limit 100;')
    lines = cursor.fetchall()
    data = dict([((a,b),c) for a,b,c in cursor.execute(
            'select * from '+args['file']+';'
            )])
    for line in lines:
        try:
            val = data[line[0],line[1]].encode('utf-8')
            print '{0}\t{1}\t{2}'.format(line[0], line[1], val) 
        except KeyError:
            pass
        
elif 'new_id' in args:
    print
    if args['new_id'] in ['new_id','newid','true']:
        cursor.execute('select DISTINCT ID from '+args['file']+';')
        linesA = [x[0] for x in cursor.fetchall()]
        cursor.execute(
            'select DISTINCT ID from backup where FILE = "'+args['file']+'";'
            )
        linesB = [x[0] for x in cursor.fetchall()]
        maxA = max(linesA)
        maxB = max(linesB)
        if maxA >= maxB:
            print str(maxA + 1)
        else:
            print str(maxB + 1)
    else:
        lines = [int(x[0]) for x in cursor.execute('select DISTINCT VAL from '+args['file']+\
                ' where COL="'+args['new_id']+'";')]
        
        print str(max(lines)+1)

elif 'file' in args:
    print 'Content-Disposition: attachment; filename="triples.tsv"'
    print
    
    # get unique columns
    if not 'columns' in args:
        cols = [line[0] for line in cursor.execute(
            'select distinct COL from '+args['file']+';')]
    else:
        cols = args['columns'].split('|')
    
    print 'ID\t'+'\t'.join(cols)
    
    # if neither concepts or doculects are passed from the args, all ids are
    # selected from the database
    if not 'concepts' in args and not 'doculects' in args:
        idxs = [line[0] for line in cursor.execute(
            'select distinct ID from '+args['file']+';')]
    else:
        # we evaluate the concept string
        cstring = 'COL = "CONCEPT" and VAL in ("'+'","'.join(args['concepts'].split('|'))+'")' if \
                'concepts' in args else ''
        dstring = 'COL = "DOCULECT" and VAL in ("'+'","'.join(args['doculects'].split('|'))+'")' if \
                'doculects' in args else ''
        
        cidxs = [line[0] for line in cursor.execute(
            'select distinct ID from '+args['file'] + ' where '+cstring)] if \
                    cstring else []
        didxs = [line[0] for line in cursor.execute(
            'select distinct ID from '+args['file'] + ' where '+dstring)] if \
                    dstring else []

        if cidxs and didxs:
            idxs = [idx for idx in cidxs if idx in didxs]
        else:
            idxs = cidxs or didxs

    # make the dictionary
    D = {}
    for a,b,c in cursor.execute('select * from '+args['file']+';'):
        if c not in ['-','']:
            try:
                D[a][b] = c.encode('utf-8')
            except KeyError:
                D[a] = {b:c.encode('utf-8')}
    # make object
    for idx in idxs:
        txt = str(idx)
        for col in cols:
            try:
                txt += '\t'+D[idx][col]
            except:
                txt += '\t'
        print txt

# XXX note that the following formula will bring much faster results:
# select * from table where ID||":"||COL in ("X:Y");
# with this formula we can limit the search space drastically, I think
# addon: it seems this is much faster, but also the fact that:

# * we print only once, XXX addon: this doesn't seem to be the case
# * we use python for sampling, and
# * we don't make many exact sqlite3 statements, this should be kept in mind,
# since it may speed up all the processes!
elif 'history' in args:
    print
    
    if 'limit' in args:
        limit = ' limit ' + args['limit']
    else:
        limit = ''
    cursor.execute(
            'select * from backup order by DATE DESC'+limit+';')
    backup = cursor.fetchall()

    # get ID restriction
    idres = '("'+'","'.join([str(line[1])+':'+line[2] for line in backup])+'")'

    # get the rest
    tables = [] 
    for line in cursor.execute(
        'select name from sqlite_master where name != "backup";'):
        tables += [line[0]]
    
    data = {}
    for table in tables:
        for a,b,c in cursor.execute(
                'select * from '+table+' where ID||":"||COL in '+idres+';'):
            try:
                data[table][a,b] = c
            except KeyError:
                data[table] = {(a,b):c}
    
    txt = ''
    for line in backup:
        try:
            new_val = data[line[0]][line[1],line[2]]
        except KeyError:
            new_val = '???'

        print '{0}\t{1}\t{2}\t{3}\t{4}\t{5}\t{6}'.format(
                line[0],
                line[1],
                line[2],
                line[3].encode('utf-8'),
                new_val.encode('utf-8'),
                str(line[4]),
                line[5])

